### PR TITLE
[fix](memory) Refactor refresh workload groups weighted memory ratio and record refresh interval memory growth

### PR DESCRIPTION
**Pull request by xinyiZzz**
1. Modify `refresh_wg_memory_info` to `refresh_wg_weighted_memory_ratio`,
- Previously, `total memory used` and `weighted memory` of wordload groups were refreshed periodically, and they no longer change during the refresh period.
- Now only `weighted memory ratio` is refreshed periodically, and `weighted memory` will use (`total memory used` + `refresh interval memory growth`) * `weighted memory ratio` in real time. Wordload groups and queries only save `weighted memory ratio`, which is more accurate.
- wordload groups `total memory used` is refreshed when `make_memory_tracker_snapshots` is executed, which is called by `workload gorups memory GC` and `refresh_wg_weighted_memory_ratio`.

2. `Thread Context` add workload group weak ptr, and record workload group memory ratio refresh interval memory growth.
- if thread in query or load, initialize with memory tracker, query id and workload group wptr.
- else not query or load, initialize with memory tracker, empty query id and default normal workload group.

### CODE REVIEW

**Comment by xinyiZzz** [comment]
run buildall

File: be/src/runtime/workload_group/workload_group.h [hunk 1]
```diff
@@ -186,7 +167,7 @@ class WorkloadGroup : public std::enable_shared_from_this<WorkloadGroup> {
     // new query can not submit
     // waiting running query to be cancelled or finish
     bool _is_shutdown = false;
-    std::unordered_map<TUniqueId, std::weak_ptr<QueryContext>> _query_ctxs;
```
Comment (yiguolei) [inline be/src/runtime/workload_group/workload_group.h:189]: 这个不要改，我们需要在wg 中保存query ctx的weak ptr。
因为我们总是需要两条路径，通过queryctx 找到 wg；从wg 找到所有的query。

**Comment by xinyiZzz** [comment]
run buildall

File: be/src/runtime/workload_group/workload_group.h (hunk 1 above)
Comment (xinyiZzz) [reply be/src/runtime/workload_group/workload_group.h]: done

File: be/src/runtime/memory/mem_tracker_limiter.h [hunk 2]
```diff
@@ -202,6 +202,18 @@ class MemTrackerLimiter final : public MemTracker {
         return querytid;
     }
 
+    void set_weighted_memory(int64_t weighted_limit, double weighted_ratio) {
+        std::lock_guard<std::mutex> l(_weighted_mem_lock);
+        _weighted_limit = weighted_limit;
```
Comment (yiguolei) [inline be/src/runtime/memory/mem_tracker_limiter.h:207]: 我感觉这两个变量还是放到query ctx 中比较好，
memtracker 只要把memory track 准就可以了，不要把这些预估的计算也放到memtracker 里。

File: be/src/runtime/memory/thread_mem_tracker_mgr.h [hunk 3]
```diff
@@ -291,6 +291,9 @@ inline bool ThreadMemTrackerMgr::try_reserve(int64_t size) {
         _limiter_tracker_raw->release(size); // rollback
         return false;
     }
+    if (_limiter_tracker_raw->wg_refresh_interval_memory_growth != nullptr) {
```
Comment (yiguolei) [inline be/src/runtime/memory/thread_mem_tracker_mgr.h:294]: 我们最终还是要在thread context 中放一个workload group的weak ptr。比如我们做IO 隔离，我们会在page io 那个地方就要获取当前在哪个wg 里，获得wg 相关的io limiter，此时也得依赖这个获得。

File: be/src/runtime/memory/mem_tracker_limiter.h (hunk 2 above)
Comment (xinyiZzz) [reply be/src/runtime/memory/mem_tracker_limiter.h:207]: done

File: be/src/runtime/memory/thread_mem_tracker_mgr.h (hunk 3 above)
Comment (xinyiZzz) [reply be/src/runtime/memory/thread_mem_tracker_mgr.h]: done

**Comment by xinyiZzz** [comment]
run buildall

File: be/src/runtime/load_stream.cpp [hunk 4]
```diff
@@ -361,18 +362,24 @@ LoadStream::LoadStream(PUniqueId load_id, LoadStreamMgr* load_stream_mgr, bool e
...
-    _query_thread_context = {load_tid, MemTrackerLimiter::create_shared(
-                                               MemTrackerLimiter::Type::LOAD,
-                                               fmt::format("(FromLoadStream)Load#Id={}",
-                                                           ((UniqueId)load_id).to_string()))};
+    _query_thread_context = {
+            load_tid,
+            MemTrackerLimiter::create_shared(
+                    MemTrackerLimiter::Type::LOAD,
+                    fmt::format("(FromLoadStream)Load#Id={}", ((UniqueId)load_id).to_string())),
+            ExecEnv::GetInstance()->workload_group_mgr()->get_task_group_by_id(1)};
```
Comment (yiguolei) [inline be/src/runtime/load_stream.cpp:382]: 这里不要穿了，直接认为是null吧

File: be/src/runtime/memory/thread_mem_tracker_mgr.h [hunk 5]
```diff
@@ -291,6 +297,10 @@ inline bool ThreadMemTrackerMgr::try_reserve(int64_t size) {
         _limiter_tracker_raw->release(size); // rollback
         return false;
     }
+    auto wg_ptr = _wg_wptr.lock();
+    if (!wg_ptr) {
+        wg_ptr->add_wg_refresh_interval_memory_growth(size);
```
Comment (yiguolei) [inline be/src/runtime/memory/thread_mem_tracker_mgr.h:302]: 要返回true or false 表示wg 是否reserve 成功了。

File: be/src/runtime/load_channel.cpp [hunk 6]
```diff
@@ -46,13 +47,16 @@ LoadChannel::LoadChannel(const UniqueId& load_id, int64_t timeout_s, bool is_hig
...
+        _query_thread_context = {_load_id.to_thrift(), query_context->query_mem_tracker,
+                                 query_context->workload_group()};
     } else {
         _query_thread_context = {
                 _load_id.to_thrift(),
                 MemTrackerLimiter::create_shared(
                         MemTrackerLimiter::Type::LOAD,
-                        fmt::format("(FromLoadChannel)Load#Id={}", _load_id.to_string()))};
+                        fmt::format("(FromLoadChannel)Load#Id={}", _load_id.to_string())),
+                ExecEnv::GetInstance()->workload_group_mgr()->get_task_group_by_id(
```
Comment (yiguolei) [inline be/src/runtime/load_channel.cpp:58]: 别-1了，直接设置为null吧

File: be/src/runtime/memory/thread_mem_tracker_mgr.h (hunk 5 above)
Comment (xinyiZzz) [reply be/src/runtime/memory/thread_mem_tracker_mgr.h]: donw

File: be/src/runtime/load_channel.cpp (hunk 6 above)
Comment (xinyiZzz) [reply be/src/runtime/load_channel.cpp]: donw

**Comment by xinyiZzz** [comment]
run buildall

**Review by yiguolei — APPROVED** [review]

**Review by wangbo — APPROVED** [review]
LGTM

**Pull request merged**